IGNITE-23440 Sql. Common internal API for cancelling an operation of a certain type

xtern opened the pull request:
https://issues.apache.org/jira/browse/IGNITE-23440

- Added `sql-engine-api` module
- Added `org.apache.ignite.internal.sql.engine.api.kill` package (naming is discussable)
- Added common interfaces:
  - OperationKillHandler - handler able to cancel operations
  - CancellableOperationType - enumeration of supported cancellable operations
  - KilllHandlerRegistry - common registry of cancel handlers (is also responsible for converting a local (node) cancel handler into a cluster-wide one).
-  Added implementation of SQL query cancellation handler
- Added integration of handlers registry with sql query execution and corresponding integration test `ItSqlKillCommandTest`

xtern commented on modules/sql-engine-api/README.md:
```diff
@@ -0,0 +1 @@
+# Apache Ignite SQL Engine API module
```
Description will be added after the first round of review.

xtern commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/message/MessageServiceImpl.java:
```diff
@@ -121,6 +122,10 @@ public void register(MessageListener lsnr, short type) {
     }
 
     private void onMessage(String consistentId, NetworkMessage msg) {
+        if (msg instanceof CancelOperationRequest) {
```
cannot use multiple message groups in a single module https://issues.apache.org/jira/browse/IGNITE-23798

korlov42 reviewed (commented):
seems this patch lacks integration with execution service: KILL statement request must be propagated to cancellation handler

xtern commented:
> seems this patch lacks integration with execution service: KILL statement request must be propagated to cancellation handler

Thanks, fixed. Added integration with SQL KILL statement.

xtern commented on modules/sql-engine-api/src/main/java/org/apache/ignite/internal/sql/common/cancel/api/CancelHandlerRegistry.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Registry of handlers that can cancel a specific operation.
+ */
+public interface CancelHandlerRegistry {
```
done

xtern commented on modules/sql-engine-api/src/main/java/org/apache/ignite/internal/sql/common/cancel/api/CancelHandlerRegistry.java:
```diff
@@ -0,0 +1,41 @@
+     * @param type Type of the cancellable operation.
+     * @return Handler that can cancel an operation of the specified type across the entire cluster.
+     */
+    OperationCancelHandler handler(CancellableOperationType type);
```
Removed from API.
Initially, I thought this registry would probably be used not only by sql-engine, but also, for example, by REST  or CLI :thinking:

xtern commented on modules/sql-engine-api/src/main/java/org/apache/ignite/internal/sql/common/cancel/api/CancelHandlerRegistry.java:
```diff
@@ -0,0 +1,41 @@
+     * @param local {@code True} if the handler can cancel operations only on local node,
+     *              {@code false} if the handler can cancel operations across the entire cluster.
+     */
+    void register(CancellableOperationType type, OperationCancelHandler handler, boolean local);
```
Done

xtern commented on modules/sql-engine-api/src/main/java/org/apache/ignite/internal/sql/common/cancel/api/OperationCancelHandler.java:
```diff
@@ -0,0 +1,41 @@
+     * @return {@code true} if the operation was successfully canceled,
+     *         {@code false} if a specific operation was not found or is inactive.
+     */
+    CompletableFuture<Boolean> cancelAsync(UUID objectId);
```
Done.

lowka commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/kill/messages/CancelOperationRequest.java:
```diff
@@ -0,0 +1,36 @@
+    String id();
+
+    /** Returns type of the operation. */
+    String type();
```
I think this should be an int that identifies an element of an enumeration because other messages tend to not use strings for enums.

xtern replied:
Done, thanks.

lowka commented on modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImplTest.java:
```diff
@@ -997,6 +1005,49 @@ public void testTimeoutDdl() {
+
+        killHandlerRegistry.register(new OperationKillHandler() {
+            @Override
+            public CompletableFuture<Boolean> cancelAsync(String operationId) {
```
I think we do not need `Thread.sleep` here and we can simply return a `CompletableFuture` that never completes.

xtern replied:
Done, thanks.

korlov42 reviewed (changes requested):
proposed code organization looks quite strange, tbh: we have registry which requires topology service for some reason; kill command is processed by execution service, but distributed coordination os done by registry

I would like to propose to organize it similar to DDL processing:
* introduce package `org.apache.ignite.internal.sql.engine.exec.kill`
    * move to this package `KillCommand` and `KillHandlerRegistryImpl`
* rename `KillHandlerRegistryImpl` to `KillComamndHandler`
    * it still implements `KillHandlerRegistry`
    * all processing is done within `KillComamndHandler`, execution service only passes the command

xtern commented on modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java:
```diff
@@ -1077,7 +1078,8 @@ public class IgniteImpl implements Ignite {
                 lowWatermark,
-                threadPoolsManager.commonScheduler()
+                threadPoolsManager.commonScheduler(),
+                new KillCommandHandler(name, logicalTopologyService, clusterSvc.messagingService())
```
The construction is left outside of QueryProcessor as I assume that other components (compute, txmanager) will use this registry to register their kill handlers.

xtern commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/message/SqlQueryMessageGroup.java:
```diff
@@ -40,4 +40,10 @@ public final class SqlQueryMessageGroup {
     public static final short QUERY_CLOSE_MESSAGE = 5;
+
+    /** See {@link CancelOperationRequest} for the details. */
+    public static final short OPERATION_CANCEL_REQUEST = 100;
```
Separate codes are used because I assumed that these messages will move to a separate message group in the future. If this assumption is incorrect, it should be changed to have continuous numbering.

xtern replied:
I thought that since they are not related to the execution of SQL queries (and are not executed through the `MessageService` introduced in the sql-engine module) it is better to put them into a separate group.
But overall it probably doesn't matter, so I added the continuous numbering

xtern commented:
> proposed code organization looks quite strange, tbh: we have registry which requires topology service for some reason; kill command is processed by execution service, but distributed coordination os done by registry
> 
> I would like to propose to organize it similar to DDL processing:
> 
>     * introduce package `org.apache.ignite.internal.sql.engine.exec.kill`
>       
>       * move to this package `KillCommand` and `KillHandlerRegistryImpl`
> 
>     * rename `KillHandlerRegistryImpl` to `KillComamndHandler`
>       
>       * it still implements `KillHandlerRegistry`
>       * all processing is done within `KillComamndHandler`, execution service only passes the command

Done.
But from my point of view it would be better to keep `HandlerRegistryImpl` where it was implemented and add a separate KillCommandHandler that will use the registry :sunglasses:  Otherwise, the area of ​​responsibility of KIllCommandHandler is not very clear now (why is it responsible for registering handlers)

xtern commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java:
```diff
@@ -72,7 +72,9 @@ public AsyncResultSetImpl(
-        return cursor.queryType() == SqlQueryType.QUERY || cursor.queryType() == SqlQueryType.EXPLAIN;
+        return cursor.queryType() == SqlQueryType.QUERY
+                || cursor.queryType() == SqlQueryType.EXPLAIN
+                || cursor.queryType() == SqlQueryType.KILL;
```
Done

xtern commented on modules/sql-engine-api/src/main/java/org/apache/ignite/internal/sql/engine/api/kill/OperationKillHandler.java:
```diff
@@ -0,0 +1,51 @@
+    /**
+     * Returns whether the handler can abort operations only on local node or across the entire cluster.
+     *
+     * @return {@code True} if the handler can abort operations only on local node,
```
Done, thanks

xtern commented on modules/sql-engine-api/src/main/java/org/apache/ignite/internal/sql/engine/api/kill/OperationKillHandler.java:
```diff
@@ -0,0 +1,51 @@
+     *
+     * @param operationId ID of the operation to cancel.
+     * @return {@code true} if the operation was successfully canceled,
+     *         {@code false} if a specific operation was not found or is inactive.
```
Removed "or is inactive".
I think the documentation will clarify the subtleties of this command's behavior. :roll_eyes:

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/fsm/Query.java:
```diff
@@ -137,4 +137,10 @@ void onError(Throwable th) {
+    CompletableFuture<Void> cancel() {
+        cancel.cancel();
+
+        return onPhaseStarted(ExecutionPhase.TERMINATED);
```
looks like cancellation via KILL and cancellation via CancelHandle completes at different moments... We need to unify it (see last `if` in `QueryExecutor#trackQuery` for details)

xtern replied:
I hope I fixed it.
I introduced new methods in Query

`register(runningQueries)` -> self registration of a query in the queries registry :thinking:
and
`cancel()` -> cancels the query and `CancellationToken` now also uses this method to cancel the query.

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/kill/LocalToClusterKillHandlerWrapper.java:
```diff
@@ -0,0 +1,149 @@
+
+    LocalToClusterKillHandlerWrapper(
+            OperationKillHandler localHandler,
+            CancellableOperationType type,
```
I think, it's better to derive type from handler itself

xtern replied:
Missed it, fixed, thanks

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/kill/LocalToClusterKillHandlerWrapper.java:
```diff
@@ -0,0 +1,149 @@
+                                        request.operationId(), request.typeId(), node.name());
+                            }
+
+                            if (response.error() != null) {
```
first, it's better to merge this `if` and previous one. Second, exception thrown from this if will trigger following `whenComplete`, resulting in warning printed to log twice.

xtern replied:
Thanks. This code was reworked. Now logging is performed only once.

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/kill/LocalToClusterKillHandlerWrapper.java:
```diff
@@ -0,0 +1,149 @@
+                        return;
+                    }
+
+                    result.completeExceptionally(throwable);
```
not sure about contract of exceptional `CompletableFuture.allOf`, but throwable will be the first thrown exception, while in fact several request may fail with different errors. I suggest to link all exception in chain of suppressed exceptions (see `org.apache.ignite.internal.sql.engine.exec.ExecutionServiceImpl#deriveExceptionFromListOfFutures` for example). It worth to add test on that matter

xtern replied:
Thanks, fixed.
Added this case in test `KillCommandHandlerTest.testCancelFinishesWithErrorOnRemoteNode`

xtern commented on modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/kill/ItSqlKillCommandTest.java:
```diff
@@ -0,0 +1,153 @@
+    @Test
+    public void killWithInvalidQueryIdentifier() {
+        assertThrowsSqlException(
+                Common.INTERNAL_ERR,
```
I hope I fixed it.
Added javadoc comment in `OperationKillHandler#cancelAsync`
```
* @throws IllegalArgumentException If the operation identifier is not in the correct format.
```
This exception is handled separately in `KIllCommandHandler`

xtern commented on modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/kill/ItSqlKillCommandTest.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    @Test
+    public void testKillQueryMetadata() {
```
Fixed, thanks
